lib/elixir/src/elixir_def.erl

```diff
@@ -129,13 +129,13 @@ head_and_definition_meta(false, _Meta, _HeadDefaults, _All) ->
 
 store_definition(Kind, {Call, Body}, Pos) ->
   E = elixir_locals:get_cached_env(Pos),
-  store_definition(Kind, false, Call, Body, E);
+  store_definition(Kind, Call, Body, E);
```

Comment:
The unquote means we have a meta-programmed function. It is meant to detect this case:

```
def foo, do: unquote(...)
```

But of course it is detecting this case:

```
def foo do
  quote do
    unquote(...)
  end
end
```

My suggestion is to improve `elixir_quote:has_unquotes` so it does not consider unquote/unquote_splicing inside a quote to be this type of unquote. But keep in mind it would have to consider this an unquote.

```
quote do: unquote(unquote(1))
```

:)

Reply:
There is also an argument that we should warn for unused functions even if it was meta-programmed in. But if we do this change, it would be in addition to the change above. :)

Reply:
Thanks a lot for the feedback! This was more tricky than I thought, but it makes perfect sense.

Out of curiosity, I tried to make an inventory of the warnings being concerned (set Check to false and checked failing tests):
1. unused function/macro/guard
2. unused default argument
3. "clauses with the same name and arity (number of arguments) should be grouped together"
4. "function head for def add/2 must come at the top of its direct implementation"

I agree that warnings for 3 and 4 could arguably be more annoying than useful when doing meta-programming.
So your strategy sounds perfect: we could warn for all the unused stuff and keep ignoring 3 and 4.

If people get warnings in existing code when we do so, they could always use `generated: true` if they need to skip it I suppose.

Adding warnings for unquotes-in-quote (except unquotes-in-unquotes-in-quotes 😁) makes sense too, although it would only concern warning 3 and 4. I think I'll start with the warn on unused part which can close the original issue and shouldn't depend on it.

Will also add tests for 3 and 4 with unquotes to guarantee we do not warn.

Reply:
Perfect. Agreed on everything.

Reply:
PR updated to do the first part ✅, will send a separate one for unquotes-in-quotes